src/java/org/apache/cassandra/db/counters/CounterLockManager.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.counters;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Interface for managing locks for CounterMutation.
+ * CounterMutation needs to ensure that each local counter is accessed by only one thread at a time.
+ * Please note that the id of the counter is an integer hash of the primary key of the row.
+ */
+public interface CounterLockManager
+{
+    boolean USE_STRIPED_COUNTER_LOCK_MANAGER = Boolean.getBoolean("cassandra.use_striped_counter_lock_manager");
+    CounterLockManager instance = USE_STRIPED_COUNTER_LOCK_MANAGER ? new StripedCounterLockManager() : new CachedCounterLockManager();
+
+    /**
+     * Handle to a lock for a particular key.
+     * Some expectations:
+     * - instances of this class are not thread-safe
+     * - it is not required that the underlying lock is reentrant
+     */
+    interface LockHandle
+    {
+        /**
+         * Try to get the lock. This method can be called at most once.
+         *
+         * @param timeout  timeout.
+         * @param timeUnit time unit.
+         * @return false in case the lock could not be acquired within the timeout.
+         * @throws InterruptedException in case the thread is interrupted while waiting for the lock.
+         */
+        boolean tryLock(long timeout, TimeUnit timeUnit) throws InterruptedException;
+
+        /**
+         * Unlock the lock if it was acquired and release the handle. This method is to be called even if the acquire method failed or even if tryLock has never been called.
+         * This method is to be called only once.
+         */
+        void release();
+    }
+
+    /**
+     * Grab locks for the given keys. The returned handles must be released by calling {@link LockHandle#release()}.
+     *
+     * @param keys list of keys, the Iterable is scanned only once in order to prevent side effects.
+     * @return a list of lock handles. The List can be iterated multiple times without side effects.
+     */
+    List<LockHandle> grabLocks(Iterable<Integer> keys);
```

Comment:
an important property of `grabLocks` is not documented here, and I think is not provided by the new cached locks manager.
Originally, the locks were returned in ascending order (wrt to the stripe id). This makes sure that no deadlock can happen.
I don't see the same property provided by the new locking mechanism. Don't we risk a deadlock here?

Reply:
you are right and in fact I added the stress with randomly ordered keys and it was deadlocking

we have two important cases to handle:
- unsorted keys
- duplicated keys

I have fixed these problems

for "unsorted keys" I added a sorting step
for "duplicated keys" I have switched to ReentrantLock (the alternative would have been to remove duplicates but then it looked weird to see grabLocks returning less elements than the original list)